org.eclipse.scout.rt.ui.html/src/main/java/org/eclipse/scout/rt/ui/html/UiSession.java

```diff
@@ -605,6 +611,52 @@ public void dispose() {
     m_sessionMetrics.sessionDestroyed(SESSION_TYPE);
   }
 
+  protected void runInModelJob(Runnable r, String name) {
+    runInModelJob(r, name, false);
+  }
+
+  protected void runInModelJob(Runnable r, String name, boolean wait) {
+    if (r == null) {
+      return;
+    }
+
+    // current thread is the model thread -> run
+    if (ModelJobs.isModelThread()) {
+      r.run();
+      return;
+    }
+
+    // current thread is NOT the model thread and the client session is active -> run in model job
+    final IClientSession clientSession = getClientSession();
+    if (clientSession != null && clientSession.isActive() && !clientSession.isStopping()) {
+      final ClientRunContext clientRunContext = ClientRunContexts.copyCurrent(true).withSession(clientSession, true);
+      IFuture<Void> modelJob = ModelJobs.schedule(
+          r::run,
+          ModelJobs.newInput(clientRunContext)
+              .withName(name)
+              .withExceptionHandling(null, false));// Propagate exception to caller (UIServlet)
+
+      // no need to wait -> return
+      if (!wait) {
+        return;
+      }
+
+      try {
+        // wait for the model job to complete
+        modelJob.awaitDone(2, TimeUnit.SECONDS);
```

Comment:
Not sure about the timeout duration. Having a timeout is a safety measure against deadlocks or forever blocked threads. But since we are in the process of disposing from outside a model job (i.e. valueUnbound), the user would not notice if the waiting time was longer. Maybe 30 or 60 seconds would be more appropriate?

Reply:
The user will not notice this delay if he closes the browser or navigates to another url in general, but he will notice something if the new url is a request to `\logout`. If the timeout is e.g. 60 seconds and there is a long running task blocking dispose, the user will wait 60s until the logout screen is displayed.